src/ol/render/canvas/ExecutorGroup.js

```diff
@@ -426,8 +426,9 @@ class ExecutorGroup {
 
   renderDeferred() {
     const deferredZIndexContexts = this.deferredZIndexContexts_;
-    for (const key in deferredZIndexContexts) {
-      deferredZIndexContexts[key].forEach((zIndexContext) => {
+    const zs = Object.keys(deferredZIndexContexts).sort(ascending);
+    for (let i = 0, ii = zs.length; i < ii; ++i) {
+      deferredZIndexContexts[zs[i]].forEach((zIndexContext) => {
```

Comment:
Sorting makes sure that styles with negative zIndex will be rendered before styles with positive zIndex.

Reply:
Do we care that a z-index of `-10` will be sorted before `-2` (and `10` before `2`)?

Reply:
Oh, thanks for spotting this. I just committed a fix.